niceman/distributions/debian.py

```diff
@@ -368,93 +365,212 @@ def _find_all_sources(self):
-        try:
-            out, _ = self._session.execute_command(
-                ['stat', '-c', '%Y', "/var/lib/dpkg/info/" + name + ".list"]
-            )
-            install_date = str(
-                pytz.utc.localize(
-                    datetime.utcfromtimestamp(float(out))))
-        except CommandError:  # file not found
-            install_date = None
-            pass
-
-        return install_date
-
-    def _get_pkg_versions_and_sources(self, name, architecture):
-        query = name if not architecture \
-            else "%s:%s" % (name, architecture)
-        out, _ = self._session.execute_command(
-            ['apt-cache', 'policy', query]
-        )
-        out = utils.to_unicode(out, "utf-8")
-        # dpkg -s uses the same output as apt-cache show pkg
-        ver = parse_apt_cache_policy_pkgs_output(out)
-        if not ver:
-            return None
-        _, ver = ver.popitem()  # Pull out first (and only) result
-        ver_dict = {}
-        for v in ver.get("versions"):
-            key = v.get("version")
-            ver_dict[key] = []
-            for s in v.get("sources"):
-                s = s["source"]
-                # If we haven't named the source yet, name it
-                if s not in self._source_line_to_name_map:
-                    # Make sure we can find the source
-                    if s not in self._all_apt_sources:
-                        lgr.warning("Cannot find source %s" % s)
-                        continue
-                    # Grab and name the source
-                    source = self._all_apt_sources[s]
-                    src_name = self._get_apt_source_name(source)
-                    source.name = src_name
-                    # Now add the source to our used sources
-                    self._apt_sources[src_name] = source
-                    # add the name for easy future lookup
-                    self._source_line_to_name_map[s] = src_name
-                # Look up and add the short name for the source
-                ver_dict[key].append(self._source_line_to_name_map[s])
-        return ver_dict
+    def _get_pkgs_arch_and_version(self, pkg_dicts):
+        # Convert package names to name:arch format
+        queries = []
+        for p in pkg_dicts:
+            # Use "dpkg -s pkg" to get the installed version and arch
+            architecture = p.get("architecture")
+            name = p.get("name")
+            queries.append(name if not architecture
+                           else "%s:%s" % (name, architecture))
+
+        # Find out how many packages we can query at once
+        max_len = max([len(q) for q in queries])
+        num_queries = max((_MAX_LEN_CMDLINE - 13) // (max_len + 1), 1)
+
+        cmd_results = []
+        # Execute dpkg -s, parse, and collate the output
+        for subqueries in (queries[pos:pos + num_queries]
+                           for pos in range(0, len(queries), num_queries)):
+            try:
```

Comment:
actually it shouldn't  be the modification to execute_command, but rather a dedicated helper (since we want to generate here), either `_session.execute_command_in_batches` or a generic `call_in_batches(self._session.execute_command, ['dpkg', '-s', None], args=...)` (do not like `_batches` so a better name is welcome!)

Reply:
The key issue is that sometimes I parse the results into arrays and sometimes into dictionaries. Meanwhile, I need to handle CommandErrors in different ways. 

So instead of creating call_in_batches(), I wrote batch_process_list(proc_func, proc_list, batch_len, start_val), which takes a processing function as an argument and iteratively calls that on the proc_list split into batch_len. The processing function is of the form f(batch, prev_value) -> result_value, and is written to take the results previous call as input (starting with start_val). 

This gave me the flexibility to change how I handle command output and exceptions.

I look forward to your review of the result (good or bad :).

Reply:
I describe the helper below...